New API version (2021-02-10) for recovery services site recovery

nilgupta6 opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [x] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [ ] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

nilgupta6 committed:
Adding old API version(2018-07-10) to new folder for 2021-20-10

nilgupta6 committed:
all changes for swagger (2021-02-10)

nilgupta6 commented:
@ravbhatnagar 

we are getting this error in lint diff

The delete operation is defined without a 200 or 204 error response implementation,please add it ---> 10 (10 instances)
we already have a 204 implemented for these delete APIs. Can we just bypass them?

Also there are 2 other errors present
2) The response is defined but without a default error response implementation.Consider adding it.' ---> 132 (132 instances)
--> our service defines error in a different format. Can we bypass this?

The response of operation:'ReplicationNetworkMappings_Get' is defined without 'systemData' ---> 45 (45 instances)
--> we are not populating this data for tracked resources APIs as our service does not have this data (last modified date etc ) saved for tracked resources. Can we bypass them?
Will I have to open suppression request for all 3 types of errors?
Please let me know how to proceed with these errors present.

FYI: @ayfathim / @vidyadharijami / @Arpitaji

nilgupta6 commented:
@ravbhatnagar 

This is the new PR. This is created as per your comments on PR: https://github.com/Azure/azure-rest-api-specs/pull/13422
Please review this. I have closed the other PR (13422)

nilgupta6 committed:
read me and custom word changes

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -198,14 +198,15 @@
             }
           }
         },
+        "x-ms-long-running-operation": true,
```
remove this

nilgupta6 replied:
done

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -6600,31 +6616,179 @@
-            "description": "Accepted"
+          }
+        },
+        "x-ms-long-running-operation": true,
```
remove this

nilgupta6 replied:
done

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -7711,6 +7934,10 @@
             "$ref": "#/definitions/A2AProtectionIntentManagedDiskInputDetails"
           }
         },
+        "vmOSManagedDisk": {
```
remove this

nilgupta6 replied:
done

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -7781,17 +8103,33 @@
             "$ref": "#/definitions/A2AVmManagedDiskInputDetails"
           }
         },
+        "autoProtectionOfDataDisk": {
```
remove this

nilgupta6 replied:
done

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -8250,6 +8588,36 @@
         }
       }
     },
+    "A2AProtectionIntentOSManagedDiskInputDetails": {
```
remove this

nilgupta6 replied:
done

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -17984,7 +19782,6 @@
       "properties": {
         "value": {
           "description": "The replication eligibility results details.",
-          "uniqueItems": false,
```
add this

nilgupta6 replied:
done

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -17998,28 +19795,23 @@
           "description": "The error status.",
-          "type": "string",
-          "readOnly": true
+          "type": "string"
```
add back the "readonly"

nilgupta6 replied:
done

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -18029,12 +19821,10 @@
       "properties": {
         "clientRequestId": {
           "description": "The client request Id.",
-          "type": "string",
```
add back readonly and unique items

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -20070,8 +21865,7 @@
         "reuseExistingNic": {
           "description": "A value indicating whether an existing NIC is allowed to be reused during failover subject to availability.",
-          "type": "boolean",
-          "default": false
```
add back the default value

nilgupta6 replied:
done

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -20083,8 +21877,7 @@
         },
         "tfoReuseExistingNic": {
           "description": "A value indicating whether an existing NIC is allowed to be reused during test failover subject to availability.",
```
dd back the default value

ayfathim commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -16848,26 +18623,6 @@
-      },
-      "x-ms-discriminator-value": "A2A"
-    },
     "RecoveryPlanA2AInput": {
```
Please get this verified by vinayak once

nilgupta6 replied:
resolving as no issue found

nilgupta6 commented:
Adding current DRI
@pilor : can you please review.

FYI: @ayfathim / @Arpitaji

nilgupta6 committed:
Ayesha's comments addressed

nilgupta6 committed:
adress PR comments from ayesha

pilor commented:
This PR has merge conflicts which is preventing the validation from running.  Please resolve them

nilgupta6 commented:
> This PR has merge conflicts which is preventing the validation from running. Please resolve them

@pilor : we have fixed merge conflicts. Can you please check.

pilor commented:
There are a lot of lintdiff failures you need to address

nilgupta6 commented:
> There are a lot of lintdiff failures you need to address

@pilor 

There are 4 types of lint errors
1) **RequiredReadOnlySystemData/R4009/ARMViolation.** (45 instances)
_Example Message:_ The response of operation:'ReplicationNetworkMappings_Get' is defined without 'systemData'
Our service is not storing data required to populate systemData. This is required for tracked APIs. We will take this change in next iterations.

2) **RequiredDefaultResponse/R4010/ARMViolation** (132 instances)
_Example Message:_ The response is defined but without a default error response implementation.Consider adding it.
Our service exposes error in a different format. This error response validation has been recently added. our service is old and is not using this error response format.

3) **DeleteOperationResponses/R4011/ARMViolation** (10 instances) 
_Example Message:_ The delete operation is defined without a 200 or 204 error response implementation, please add it.
In all the flagged instances we already have implemented 204.
It looks like ARM rule want us to implement all 3 (200/202/204). 200 and 202 cannot go together. This looks like an issue with the rule validations.

4)   **PathResourceTypeNameCamelCase/R3021/ARM violation** (96 instances)
_Example Message:_ "Resource type naming must follow camel case.
we have introduced new API version (2021-02-10) in our service and we have fixed all camel case issues in that versions service.json.
All camel case issues are coming from previous version (2018-07-10). We have not done any changes in previous versions service.json. Not sure why lint errors are checked in previous version.

**SDK violations**
1) we have removed few classes and have introduced new classes.
2) We have modified  one object.
3) Also we have removed one property from one class.

We have fixed all these in our new SDK.

**I have raised suppression for all these issues. Please find word doc with screenshots for suppression forms.**
https://microsoft-my.sharepoint.com/:w:/p/nilgupta/EYN6S7KLqTZIsARvHglskqoBGtqXdHVlPqsbvRWhIS5MXQ?e=grXJcA


FYI @ayfathim / @Arpitaji

nilgupta6 commented:
@pilor 
Can you please start the review.
We have raised the suppression requests. This PR is time critical for us given the external deadlines.

pilor commented:
For systemData you get one api-version of leniency but please set aside the time to implement it next time you rev your api-version now that you are aware of the requirement.

For async deletes I am OK if the final operationResults response is a 204 instead of 200.  The RPC does not require that it is 200.  You are fine suppressing that imo.

As you pointed out the camelcase errors are from an older spec.  You shouldn't have to suppress those.  Please work the sdk reviewer to understand why the linter is picking that up.

pilor commented:
The [error response contract is standard](https://github.com/Azure/azure-rest-api-specs/pull/13479) and must be followed by all RPs.  What does your error response look like and why is it different?  And why is it not defined in your swagger spec?

nilgupta6 committed:
Adding suppressions

nilgupta6 commented:
> For systemData you get one api-version of leniency but please set aside the time to implement it next time you rev your api-version now that you are aware of the requirement.
> 
> For async deletes I am OK if the final operationResults response is a 204 instead of 200. The RPC does not require that it is 200. You are fine suppressing that imo.
> 
> As you pointed out the camelcase errors are from an older spec. You shouldn't have to suppress those. Please work the sdk reviewer to understand why the linter is picking that up.

@pilor 

I have added suppression for 
RequiredReadOnlySystemData (R4009) and DeleteOperationResponses (R4011)

nilgupta6 commented:
> The [error response contract is standard](https://github.com/Azure/azure-rest-api-specs/pull/13479) and must be followed by all RPs. What does your error response look like and why is it different? And why is it not defined in your swagger spec?

For RequiredDefaultResponse (R4010)
This is a new rule by ARM and ours is a old service.
our error responses look like.

{
  "serviceErrorDetails": {
    "code": "",
    "message": "",
    "possibleCauses": "",
    "recommendedAction": "",
    "activityId": ""
  },
  "providerErrorDetails": {
    "errorCode": "",
    "errorMessage": "",
    "errorId": "",
    "possibleCauses": "",
    "recommendedAction": ""
  },
  "errorLevel": "Error",
  "creationTime": "2021-03-25T07:30:05.9364016Z",
  "taskId": "UpdateReplicationAgentTask"
}

This error states to put error response in swagger. We will work to get this in our swagger from next time.
We have raised suppression for this and it is approved
![image](https://user-images.githubusercontent.com/78136736/112524748-30668c00-8dc6-11eb-87e0-12a11e5b138b.png)

nilgupta6 commented:
@pilor 
I see that you have added the tag: **armSignedOff**. 
Can you please approve it. We dont want to merge now and will get it merged after service release.

pilor commented:
> @pilor
> I see that you have added the tag: **armSignedOff**.
> Can you please approve it. We dont want to merge now and will get it merged after service release.

Approve what?  The tag is an approval from ARM API review.  @ArcturusZhang (the sdk reviewer) will handle the rest of the SDK review + merge (if desired)

nilgupta6 commented:
> > @pilor
> > I see that you have added the tag: **armSignedOff**.
> > Can you please approve it. We dont want to merge now and will get it merged after service release.
> 
> Approve what? The tag is an approval from ARM API review. @ArcturusZhang (the sdk reviewer) will handle the rest of the SDK review + merge (if desired)

Thanks. Will work with @ArcturusZhang further.

nilgupta6 commented:
@ArcturusZhang 
We are done with ARM review. Can you please start your review.

ArcturusZhang reviewed (approved):
Approve for `readme.go.md`

nilgupta6 commented:
@lirenhe : We are done with ARM review. Can you please start your review.

lirenhe commented:
> @lirenhe : We are done with ARM review. Can you please start your review.

I found a couple of operations are removed in your latest  API version, are those APIs deprecated?

lirenhe commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -4138,175 +4319,9 @@
-          }
-        }
-      },
-      "put": {
```
Is this API removed?

nilgupta6 replied:
@lirenhe 
Can you please copy/paste some more context here (more line). Which API is this?
Github is not showing the diff as file is too large.
but we have not removed any API from swagger. We have introduced new version and have added few properties in some APIs

nilgupta6 replied:
@lirenhe : Any updates on this?

lirenhe commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2021-02-10/service.json:
```diff
@@ -4138,175 +4319,9 @@
-      }
-    },
-    "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{resourceName}/replicationFabrics/{fabricName}/replicationProtectionContainers/{protectionContainerName}/replicationProtectionContainerMappings/{mappingName}": {
-      "get": {
```
Same above, as I know, there is a review process for the deprecated API.

nilgupta6 replied:
I am seeing this API being present in swagger doc.
can you please check. It might have been added at some other location.
@lirenhe

nilgupta6 replied:
@lirenhe : Any updates on this?

nilgupta6 replied:
@lirenhe : Can you please check this..
I have updated all review comments.

lirenhe commented:
@nilgupta6, could you also do a code sync with master and then submit PR again?

nilgupta6 committed:
fixing custom words

nilgupta6 commented:
> @nilgupta6, could you also do a code sync with master and then submit PR again?

@lirenhe : I have done sync with master and updated PR.
Can you please review

nilgupta6 commented:
@lirenhe : can you please merge this PR.

FYI
@ayfathim / @Arpitaji

nschonni commented on custom-words.txt:
```diff
@@ -2139,3 +2139,8 @@ ziplist
+Orignal
+Onprem
+targetv
+Seleted
```
This is a typo, not something to ignore

nschonni replied:
Opened https://github.com/Azure/azure-rest-api-specs/pull/14120 to suppress for the single version and flag again on the next

nschonni commented on custom-words.txt:
```diff
@@ -2139,3 +2139,8 @@ ziplist
 zoneinfo
 zset
+testfailover
+Orignal
```
This is a typo, not something to ignore
